### DISCUSSION TITLE
v1.0.0

### DISCUSSION THREAD

**Discussion by cmseguin** [cmseguin/rollup-plugin-react-scoped-css#52] [Ideas]
This is a thread for all the ideas, concerns, etc... regarding the 1.0.0 release coming up soon.

So far what is up for discussion is:
+ Removing `@vue/component-compiler-utils` as a dependency
+ Change the scoping attribute away from the Vue way ie. (data-v-{hash})
+ Add dependabot
+ Add ability to have any filename scoped and only .global.scss files not scoped.

**Comment by wirdehall** [comment]
So, I don't think removing `@vue/component-compiler-utils` as a dependency needs more discussion, unless someone else wants to add something, feels like we are in agreement. 
If anyone is intrested in motivations here is the inital discussion and motivation in my original PR that pushed for a 1.0.0 release:
https://github.com/cmseguin/rollup-plugin-react-scoped-css/pull/48

Regarding having a prefix in the data-attriburte I think that should be optional. Default behaviour should be data-{hash}.
Angular does data-{hash}, if it works for them I don't see why we would need the prefix. The odds of there ever being a data attribute with the name of the hash for the specific file is astronomical and super unlikely, it's not something a human would do. It whould have to be another repo in which case if there is a conflict, having the option of adding a prefix makes sense. But I don't think it should be there by default.

**Reply by cmseguin** [reply]
How about by default it's data-{hash} with the ability to add a prefix if wanted? Someone might be running angular and react (react inside angular and might want safety between the two however unlikely there might be a collision?

**Reply by wirdehall** [reply]
Yeah, that was my proposal.

**Reply by cmseguin** [reply]
I'll add this now! :)

**Comment by wirdehall** [comment]
One more thing that could be up for fixing is all the linting warnings we are getting in the AST files.
We are currently getting 44 eslint errors when running `npm run lint` all of them warning that we are using `any`.
I haven't taken a look at that code more that just quickly scaning it for what it was doing so I'm not familliar with it enough to know how big of a undertaking that would be but it seams like something we would want to do at somepoint.

**Reply by cmseguin** [reply]
Yea I agree, this code is a little messy, I didn't spend enough time to fix these issue!

**Comment by wirdehall** [comment]
We might want to add something in the documentation about that if you are not using vite and using it with just rollup and are using non-standard CSS like SCSS or Less you will need a plugin that transpiles it down to CSS.
Vite does this for us but with just vanilla rollup you would have to bring your own transpiler.

I don't know how usual it is for people to just use rollup with this plugin but I can imagine this comming up as an Issue quickly after release of 1.0.0 if we don't mention it in the documentation.

**Reply by cmseguin** [reply]
I will be very honest, I have not tested yet with rollup only (without Vite) so I don't know how good this plugin is without Vite. I took this plugin name because I thought it would be a good idea to eventually support rollup alone. I would have to do more testing with that. But technically speaking, if the Scss plugin runs before the scoping, it should work. I will eventually try to test that, I just don't have that much time these days.

But that's a good Idea I will probably add a disclaimer somewhere.

**Reply by wirdehall** [reply]
Yeah, it should definetly work, but it might be good to mention that **you do need to add a plugin to transpile down to css** if you are using vanilla rollup without vite.
So people don't expect this plugin to do that.
I think there are better libs out there for the task of transpiling and we should focus on doing one thing and doing that good.

**Reply by cmseguin** [reply]
I will add something in the 1.0.0-alpha branch for that! :)

**Comment by wirdehall** [comment]
I can see you added requirements in the docs:

>     node >= 20.0.0
>     vite >= 4.0.0

I have tried this on node v 18.18 and it works so I don't think node v20 is a requirement.
Also, do we no longer support just the usage of rollup if vite >= 4.0 is required?
Maybe a base-requirement of node version and a requirement section for if you are using it with vite might be resonable?

**Reply by cmseguin** [reply]
yea that would be a good idea to test lower version of node to see which one it's compatible with, I think I added that because one of the dependency I added was asking for node 20 but it might have been a dev dependency... 
For rollup, yea it should but I am just not sure of any of the specific concerning the supported versions since I don't really.

**Reply by wirdehall** [reply]
Regarding rollup, same here, no clue!
But I can confirm node v18.18 works.
I just tried running the tests in node version v14.17.5, that also worked.
My current project is running on node version 18.18  so I know everything works on that version at least.

**Reply by cmseguin** [reply]
I tested on 16 and it also seems to work so I will probably go with that since that's what Vite is recommending. Updating that now! :)

**Reply by cmseguin** [reply]
I updated this to 18 since vite 5 is also deprecating 16

**Comment by wirdehall** [comment]
You know what I would want as well in this release?
For my project, all besides 3 SCSS files are scoped. That's hundreds of files where I have to use the .scoped.scss ending.
I would like to add an option in the plugin for doing the inverse.
Declare global SCSS files and assume all else are scoped.

So let's say:
While initiating the plugin
```reactScopedCssPlugin({ styleFileSuffix: 'global', defaultMode: 'scoped' })```
Where defaultMode is a new option.
WIth this config. All scss/less/css files which is missing a sufix will be scoped and only the files that has the `.global.scss` suffix will be global. 

Typedefinition:
```defaultMode?: 'scoped' | 'global'```
To perserve backwards compability if not set it will default to 'global' which will behave like it does today where you have to use the surfix for it to be scoped.

What do you think?

**Reply by wirdehall** [reply]
I know I could achive this today by setting styleFileSuffix to ```undefined``` and then add my 3 files I want to be global to the exluded list but I think it's way more obvious what's happening with the global suffix and it being automatic.

**Reply by wirdehall** [reply]
Actually, I just tried this, there appear to be some logical errors when using the `styleFileSuffix` option.
It tries to load my node_module imported css files into the plugin as well, and as if that wasn't enough I can't exclude them as the exclude filter only appears to apply to the post-processor. The pre-processor is throwing errors as the paths are not calcualte correctly for those stylesheets.

I think this one would benefit a discussion on what the most appropriate solution to this would be before I start implementing a solution.
At the very least I think the exclude filter shoudl be run in both the pre and post functions so you have the options of excluding files that are being rowdy.
What are your thoughts?

**Reply by wirdehall** [reply]
Hum... I'm getting all sorts of weird errors in react as well when using `styleFileSuffix: undefined`.
Are we sure this has been tested and works correctly?

**Reply by cmseguin** [reply]
Yea I have not really tested that part that much... It's very possible that it's not working! (my bad, this is my first open-source experience and I didn't think it would become serious lol). That being said I think this is a great idea, I would still make the default .scoped but if we can set it in some ways to do the opposite (.global for global stuff) that would be amazing! I think I might want to get rid of `styleFileSuffix` & `styleFileExtensions` in favour of some sort of include & exclude glob pattern or regex though.

default could be:
```
reactScopedCssPlugin()

// default option { include: ["*.scoped.(scss|css|sass|less)"] }
```

without .scoped could be this:
```
reactScopedCssPlugin({ include: ["*.(scss|css|sass|less)"], exclude: ["*.global.(scss|css|sass|less)"] })
```

Maybe it's not include or exclude, because I think those were also meant for jsx/tsx files, I'm not sure. But I think that would be less confusing than giving 5 options to control how files are processed. We can explore further but I agree it would be nice to have this feature in 1.0.0.

**Reply by wirdehall** [reply]
I will think about what config solution might be best. I think include or exclude is still good to have as a option for advanced management. I also think scoped/global should be a boolean as to be a simple switch of behaviour but we can explore this a bit more no hurry.

**Reply by cmseguin** [reply]
I implemented something that made sense in my head, check it out and feel free to change it a bit but I think it is pretty much what you had in mind :) ended up being the simplest way to go about it. I tested it quickly but I encourage you to test it too.

**Reply by wirdehall** [reply]
Finally got around to test this.
Scoped by default does not seam to work so good.

I'm running in to two 2 issues.
1. None of the html tags is getting the scoped data-attribute which is weird since they should always be taged by the data-attribute. I can't understand the logic in `rollup-plugin-react-scoped-css:post` anymore but I think there something wrong around this code: 
```
const matches = code.match(scopedCssInFileRegex);

const shouldScope = matches?.some((match) => {
   return !(scopeStyleByDefault && globalCssRegex.test(match));
});
```

2. This was an issue previously but the exclude filter does not seam to work since i'm still getting errors from this action: `rollup-plugin-react-scoped-css:pre`. Like this:

```
18:52:22 [vite] Pre-transform error: Failed to load url /src/styles/@fontsource/roboto/300.css?scope=1eb364ec (resolved id: /src/styles/@fontsource/roboto/300.css?scope=1eb364ec) in /path/to/project/src/styles/shared-imports.tsx. Does the file exist?
18:52:22 [vite] Error when evaluating SSR module /src/styles/shared-imports.tsx: failed to import "/src/styles/@fontsource/roboto/300.css?scope=1eb364ec"
|- Error: Failed to load url /src/styles/@fontsource/roboto/300.css?scope=1eb364ec (resolved id: /src/styles/@fontsource/roboto/300.css?scope=1eb364ec) in /path/to/project/src/styles/shared-imports.tsx. Does the file exist?
    at loadAndTransform (file:///path/to/project/node_modules/vite/dist/node/chunks/dep-V3BH7oO1.js:49731:21)
```

This is something importing from node_module like this: `import '@fontsource/roboto/300.css';`.
And I have tried more or less any and all variants of exclude but cant get it to work.
I tried just hacking in some debugging after the `if(!importer)` statement like so:
```
if((exclude as Array<string>).find((e) => source.includes(e))) {
  return;
}
```

And it worked, but that code needs to be cleaned up and obviously handle all input cases of exclude. I just handle one of 3 options there.

Anyways, I think these two things are the last things blocking version 1.0.0 from release.
Or did you get these things working? Am I doing something wrong?

Lastly, I think if we get these two issues I think it would make a lot of sense to have `scopeStyleByDefault` set to true by default.
If that's what we like it's a good time to do that change now as we are upgrading major version and still having the old way of doing it supported.

Cheers!

**Reply by wirdehall** [reply]
@cmseguin Just wanted to see if you saw this?

**Reply by cmseguin** [reply]
Sorry life got a hold of me again! I will try to fix this ASAP. Hopefully this is the last little things. 

Also concerning scopeStyleByDefault, I think I would like to gather a little more feedback from the users before changing the default behaviour. I am definitely open to the ideas if a majority of the current users wants it.

**Reply by wirdehall** [reply]
No worries I get it.
Just wanted to make sure you had seen it. It's not urgent so if life is a lot right now, don't stress.

**Reply by cmseguin** [reply]
Sorry I just got a little bit of free time..  I am pretty confident I fixed issue 1. I now parse all the jsx files and use AST to figure out the imports now. it seems to work pretty good as far as I can tell with some quick tests.

for issue 2, I am not sure, I cannot replicate, I installed this package on the example in the repo and it does work fine. Maybe I am doing something wrong?

**Comment by wirdehall** [comment]
Just writing it down here real quick as I need to go.
There is a bug in the current implementation of for scoping css.
We shouldn't append the tag after pseudo-elements like `::before` or `:not`. But we should append them before them in a chain.
Here are actuall examples from my code-base:
`.tire[data-991b57c1] .eu-labels[data-991b57c1] .eu-label[data-991b57c1] span:not(.skeleton-image):not(.extra)[data-991b57c1]::before[data-991b57c1]`
I would expect this:
`.tire[data-991b57c1] .eu-labels[data-991b57c1] .eu-label[data-991b57c1] span[data-991b57c1]:not(.skeleton-image):not(.extra)::before`

It's adding it both after `:not` and after `::before` but not after the last `span`.
If you don't have time to look at this I will create a PR tomorrow :)

**Reply by cmseguin** [reply]
I fixed that I think but I have a feeling there might be weird edge cases sometimes. We'll probably have to add tests for most pseudo cases at some points...

**Reply by wirdehall** [reply]
I will take a look at the tests and the feature now and create a PR when it's done.

**Comment by cmseguin** [comment]
I know this is not very crucial to this plugin but I would also like to add a section in the readme to acknowledge your contribution to this project. (and the other people who did PR in the past) I am not sure of the best way to do it, maybe there are automated ways for future contributors (or it needs to be manual?)

**Reply by wirdehall** [reply]
I havn't been ivolved this much in a open source project either. Only a PR here and there so I don't know.
I know Github has a feature where you can see contributors on the right on the projects "start-page" below releases and packages.

I think some projects I have seen have had a dedicated thanks to some of the more involved contributors but I don't think there is any standard for this that I have seen.

**Comment by wirdehall** [comment]
We should add dependabot for automated PR's for outdated deps as well :)

**Reply by cmseguin** [reply]
Good idea! I will add to the list!

**Comment by cmseguin** [comment]
So I found a pretty big limitation with this plugin... It has to do with scoping sub components, and I am not 100% sure there is a lot to do to fix it.

The problem is with a scoped sub component that has classes let's say the following:

```
// Inner.tsx
const Inner = ({ className }) => {
  return (
    <h1 className={`internal ${className}`}>heading</h1>
  )
}

// Outer.tsx
const Outer = () => {
  return (
    <div>
      <Inner className="external" />
    </div>
  )
}
```

It will compile the jsx to this:

```
<div data-123>
    <h1 data-456 className="internal external">heading</h1>
</div>
```

However to properly style this element you might expect this to compile to that instead:
```
<div data-123>
    <h1 data-456 data-123 className="internal external">heading</h1>
</div>
```
This would allow you to scope the inner component itself too.

the only way to achieve this currently without modifying this plugin with very convoluted code that will be prone to a lot of bugs is to add {...props} to the outer element of the Inner component (h1). But as you might imagine this is very limited. 

Not sure the best solution, food for thought...

**Reply by wirdehall** [reply]
I don't think this is something you should do or something the plugin should handle.
If you look at Angular and Vue's scoping of css, this is not how they handle it.

You scope only for the current component and if you want to apply CSS to nested components you apply the ::deep pseudo-element in your css.

The correct response from our plugin would be:
```
<div data-123>
    <h1 data-456 className="internal external">heading</h1>
</div>
```

**Reply by wirdehall** [reply]
I.e. if you wanted to style the inner component from the outer componenet you would do something like this:

```
// Outer.css
div::deep > .external {
  // Your css here
}
```

**Reply by cmseguin** [reply]
That makes sense! Yea I could achieve that but I might just update the docs because as I explore the ins and out of scoping I find new ways of using it that might not be good or as intended.

**Comment by wirdehall** [comment]
I found an issue, tried to upgrade to upgrade to Vite v5.0.2 which was released last week.
It did not work, apparently they upgraded to rollup v4 which means they no longer support Acorn plugins.
I have not had the time to research what that is but apparently our plugin is a Acorn plugin.

Hopefully this is something easy and quick to fix.

**Reply by cmseguin** [reply]
Yea hopefully I can just upgrade a couple of things. I will take a look at that when I get a chance! :)

**Reply by cmseguin** [reply]
It should work with vite 5 now but I guess I will deprecate vite 4 since it now depends on rollup 4. It might work still but I feel like it's just safer to say we support only vite 5.

**Reply by wirdehall** [reply]
Sounds good!
Will test it as soon as I have time.

**Reply by wirdehall** [reply]
I have tested and verifed that vite 5.0.11 now works flawlessly with this plugin. Good job!

**Comment by wirdehall** [comment]
On another positive note, it feels like we are kind of stable as of right now. If we fix the Acorn issue we should be good to go. I haven't found any issue for the last two weeks with using this plugin. The data tags is applied correctly and the css scoping works flawlessly.
We might see a 1.0.0 release not too far in the future, or what do you think? :)

**Reply by cmseguin** [reply]
I agree the styling works as expected from what I have been working with! I would like to figure out the part where we can scope files without a prefix but after that I think it will be ready to promote. I think we can skip beta since no one else is really testing this anyway lol.

**Reply by cmseguin** [reply]
I changed a few things tonight, I feel like we should test that latest alpha version, however code wise I think I am happy with that. I would like to maybe polish the readme before merging the alpha but I think we are really close.

**Reply by wirdehall** [reply]
Will test as soon as I have some time.

**Comment by wirdehall** [comment]
Just installed [1.0.0-alpha.16](https://www.npmjs.com/package/rollup-plugin-react-scoped-css/v/1.0.0-alpha.16) on a completely new vite project and I'm getting this typescript error on the line where I add the plugin using the function `reactScopedCssPlugin()`:

```
No overload matches this call.
  The last overload gave the following error.
    Type 'VitePartialPlugin[]' is not assignable to type 'PluginOption'.
      Type 'VitePartialPlugin[]' is not assignable to type 'PluginOption[]'.
        Type 'VitePartialPlugin' is not assignable to type 'PluginOption'.
          Type 'VitePartialPlugin' is not assignable to type 'Plugin<any>'.
            Types of property 'load' are incompatible.
              Type 'ObjectHook<(this: PluginContext, id: string) => LoadResult | Promise<LoadResult>, {}>' is not assignable to type 'ObjectHook<(this: PluginContext, id: string, options?: { ssr?: boolean; }) => LoadResult | Promise<LoadResult>>'.
                Type '(this: PluginContext, id: string) => LoadResult | Promise<LoadResult>' is not assignable to type 'ObjectHook<(this: PluginContext, id: string, options?: { ssr?: boolean; }) => LoadResult | Promise<LoadResult>>'.
                  Type '(this: PluginContext, id: string) => LoadResult | Promise<LoadResult>' is not assignable to type '(this: PluginContext, id: string, options?: { ssr?: boolean; }) => LoadResult | Promise<LoadResult>'.
                    Type 'import("/path/to/project/node_modules/rollup-plugin-react-scoped-css/node_modules/rollup/dist/rollup").LoadResult | Promise<import("/path/to/project/node_modules/rollup-plugin-react-scoped-css/node_modules/rollup/dist/rollup").LoadResult>' is not assignable to type 'import("/path/to/project/node_modules/rollup/dist/rollup").LoadResult | Promise<import("/path/to/project/node_modules/rollup/dist/rollup").LoadResult>'.
                      Type 'SourceDescription' is not assignable to type 'LoadResult | Promise<LoadResult>'.
                        Type 'import("/path/to/project/node_modules/rollup-plugin-react-scoped-css/node_modules/rollup/dist/rollup").SourceDescription' is not assignable to type 'import("/path/to/project/node_modules/rollup/dist/rollup").SourceDescription'.
                          Types of property 'ast' are incompatible.
                            Type 'AstNode' is not assignable to type 'ProgramNode'.
                              Type 'AstNode' is missing the following properties from type 'Program': sourceType, bodyts(2769)
index.d.ts(3032, 18): The last overload is declared here.
```
This is my entire `vite.config.ts`:

```
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react-swc'
import { reactScopedCssPlugin } from 'rollup-plugin-react-scoped-css';

export default defineConfig({
  plugins: [
    react(), 
    reactScopedCssPlugin(),
  ],
})
```
I think we are supposed to return `PluginOption[]` instead of `VitePartialPlugin[]` but you might know better :)

**Reply by cmseguin** [reply]
VitePartialPlugin extends Plugin from rollup normally the types should intersect especially for that load function.. it only adds a little option but that's it really. 

What are you using to install vite or what are you using to install your deps. Is it possible that you are using yarn > 2 or pnpm? They try to memoize the packages and it's possible it tried to use the wrong version of rollup. On my end I have the most recent version of the plugin and vite and it seems to work correctly. But I would be curious, maybe I am doing something wrong! :)

**Reply by wirdehall** [reply]
Hum... I had an error for like two days.
And then, poff, it dissapeared.
I'm super confused.

It appears to work now. Maybe there were some issue with VsCode or something? I don't know 😅

**Comment by wirdehall** [comment]
Hum... we should probably try to get this finished at some point and get it released.
Are you waiting for me on something or what's the current status?

Did we ever solve this? I don't even remeber anymore since it's been so long: https://github.com/cmseguin/rollup-plugin-react-scoped-css/discussions/52#discussioncomment-7501086

Not sure this was ever fixed: https://github.com/cmseguin/rollup-plugin-react-scoped-css/discussions/52#discussioncomment-8374729

I can say that Alpha 16 is working great in many aspects.
I am running it on multiple projects and as long as I don't try scopeStyleByDefault or use "exception patterns" it works great.
